src/server/pkg/sync/sync.go

```diff
@@ -179,20 +196,56 @@ func (p *Puller) Pull(client *pachclient.APIClient, root string, repo, commit, f
 			})
 		}
 		if emptyFiles {
-			return p.makeFile(path, func(w io.Writer) error { return nil })
+			_, err := p.makeFile(path, false, func(w io.Writer) error { return nil })
+			return err
 		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	limiter := limit.New(concurrency)
+	var eg errgroup.Group
+	if !pipes && !emptyFiles {
 		eg.Go(func() (retErr error) {
 			limiter.Acquire()
 			defer limiter.Release()
-			return p.makeFile(path, func(w io.Writer) error {
-				return client.GetFile(repo, commit, fileInfo.File.Path, 0, 0, w)
-			})
+			var oldFile *os.File
+			// for the common case, when neither pipes nor emptyFiles are set, we make a recursive
+			// GetFiles call in order to retrieve all the data for the repo from object storage in a single call.
+			if err := client.GetFiles(repo, commit, file, 0, 0, func(fileInfo *pfs.FileInfo, r io.Reader) error {
+				if fileInfo != nil && fileInfo.File != nil {
```

Comment:
Is there ever a valid scenario where fileInfo is non-nil and fileInfo.File is nil? If not, then I think we should maybe think about refactoring some of this because we could let an error slip by unnoticed. I think it might make more sense to have this just check for fileInfo being nil and then handle the potential fileInfo.File being nil within the block. Also, I think it might be cleaner to have the make file callback be a no-op, and then take the copy out of the else block.

Reply:
Nope, now that `getFiles()` generates the fileInfo using the `nodeToFileInfo()` method we are guaranteed to have a non-nil File for every non-nil fileInfo. This was leftover from before I was using `nodeToFileInfo()` to properly handle the header/footer cases. Fixed.